poptorch_experimental_addons/_impl/collectives.py

```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import torch
+import poptorch
+
+
+def replicated_all_gather(x: torch.Tensor, replication_factor: int) -> torch.Tensor:
+    """
+    SPMD all-gather across IPUs.
+
+    Gathers and concatenates tensors occupying the same memory location across IPUs,
+    then replicates the result.
+    """
+    out = poptorch.custom_op(
+        [x],
+        name="ReplicatedAllGather",
+        domain="ai.graphcore",
+        domain_version=1,
+        example_outputs=[torch.randn(replication_factor, *x.shape)],
+    )[0]
+    out = out.reshape(replication_factor * x.shape[0], *x.shape[1:])
```

Comment:
I would expect the output to have rank `rank(x) + 1`, a leading axis of the `replication_factor` for general utility, but happy to be disagreed with.

Reply:
I'll check this, although I assumed it should act as a `concatenate`, rather than a `stack`

Reply:
After checking https://pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather (which returns a list of tensors), I agree with you that this should return an object with length `replication_factor`, but it makes more sense to us for this object to be a tensor.